services/src/kilda-core/kilda-persistence-neo4j/src/main/java/org/openkilda/persistence/repositories/impl/Neo4jIslRepository.java

```diff
@@ -0,0 +1,172 @@
+            throw new PersistenceException(format("Found more that 1 ISL entity with %s_%d - %s_%d",
+                    srcSwitchId, srcPort, dstSwitchId, dstPort));
+        }
+        return isls.isEmpty() ? null : isls.iterator().next();
+    }
+
+    @Override
+    public Collection<Isl> findActiveAndOccupiedByFlowWithAvailableBandwidth(String flowId, long requiredBandwidth) {
+        Map<String, Object> parameters = ImmutableMap.of(
+                "flow_id", flowId,
+                "requested_bandwidth", requiredBandwidth,
+                "switch_status", switchStatusConverter.toGraphProperty(SwitchStatus.ACTIVE),
+                "isl_status", islStatusConverter.toGraphProperty(IslStatus.ACTIVE));
+
+        String query = "MATCH (src:switch)-[fs:flow_segment{flowid: $flow_id}]->(dst:switch) "
+                + "MATCH (src)-[link:isl]->(dst) "
+                + "WHERE src.state = $switch_status AND dst.state = $switch_status AND link.status = $isl_status "
+                + " AND link.src_port = fs.src_port AND link.dst_port = fs.dst_port "
+                + " AND link.available_bandwidth + fs.bandwidth >= $requested_bandwidth "
+                + "RETURN src, link, dst";
+
+        return Lists.newArrayList(getSession().query(getEntityType(), query, parameters));
+    }
+
+    @Override
+    public Collection<Isl> findAllActive() {
+        // 0 bandwidth means ignore it.
+        return findActiveWithAvailableBandwidth(0L);
+    }
+
+    @Override
+    public Collection<Isl> findActiveWithAvailableBandwidth(long requiredBandwidth) {
+        Map<String, Object> parameters = ImmutableMap.of(
+                "requested_bandwidth", requiredBandwidth,
+                "switch_status", switchStatusConverter.toGraphProperty(SwitchStatus.ACTIVE),
+                "isl_status", islStatusConverter.toGraphProperty(IslStatus.ACTIVE));
+
+        String query = "MATCH (src:switch)-[link:isl]->(dst:switch) "
+                + "WHERE src.state = $switch_status AND dst.state = $switch_status AND link.status = $isl_status "
+                + " AND link.available_bandwidth >= $requested_bandwidth "
+                + "RETURN src, link, dst";
+
+        return Lists.newArrayList(getSession().query(getEntityType(), query, parameters));
+    }
+
+    @Override
+    public void updateIslAvailableBandwidth(SwitchId srcSwitchId, int srcPort, SwitchId dstSwitchId, int dstPort) {
+        Map<String, Object> parameters = ImmutableMap.of(
+                SRC_SWITCH_PROPERTY_NAME, srcSwitchId.toString(),
+                SRC_PORT_PROPERTY_NAME, srcPort,
+                DST_SWITCH_PROPERTY_NAME, dstSwitchId.toString(),
+                DST_PORT_PROPERTY_NAME, dstPort);
+
+        String query = "MATCH (src:switch {name: $src_switch}), (dst:switch {name: $dst_switch}) "
+                + "WITH src,dst "
+                + "MATCH (src) - [i:isl { src_port: $src_port, dst_port: $dst_port }] -> (dst) "
+                + "WITH src,dst,i "
```

Comment:
this `with` and previous are excess... Same as fist `match`... 
```
MATCH (src:switch {name: $src_switch})) - [i:isl { src_port: $src_port, dst_port: $dst_port }] -> (dst:switch {name: $dst_switch})
"OPTIONAL MATCH (src) - [fs:flow_segment { ...
```

Reply:
Refactored to 
```
"MATCH (src:switch {name: $src_switch}), (dst:switch {name: $dst_switch}) "
+ "WITH src,dst "
+ "OPTIONAL MATCH (src) - [fs:flow_segment { "
+ " src_port: $src_port, "
+ " dst_port: $dst_port, "
+ " ignore_bandwidth: false "
+ "}] -> (dst) "
+ "WITH sum(fs.bandwidth) AS used_bandwidth RETURN used_bandwidth"
```